### PR TITLE
Improving stability, board support for ULX3S

### PR DESCRIPTION
**Pull request by emard**
Here is improved python application with much stability
because it will retry 10 each failed flash sector before giving up.

It has been tested on ULX3S (so here's also this board support).
It can generate multiboot image from ddtcmd tool, but I don't know how
to make user image be replaceable by later flashing with tinyprog.

tinyprog will flash it properly but replaced image won't be loaded at pulling
PROGRAMN to GND.

### CODE REVIEW

File: programmer/tinyprog/__init__.py [hunk 1]
```diff
@@ -386,6 +467,60 @@ def boot(self):
             # for the boot command data packet.
             pass
 
+    # similar to program_sectors
```
Comment (tinyfpga) [inline programmer/tinyprog/__init__.py]: This is almost exactly the same as `program_sectors`...there is way too much duplication of code here.

File: programmer/tinyprog/__init__.py [hunk 2]
```diff
@@ -386,6 +467,60 @@ def boot(self):
...
+    def verify_sectors(self, addr, data, retry=10):
```
Comment (tinyfpga) [inline programmer/tinyprog/__init__.py]: This function has  too much copy and pasted code.  Most of the code that was pasted in this function is not used.

File: programmer/tinyprog/__init__.py [hunk 3]
```diff
@@ -386,6 +467,60 @@ def boot(self):
...
+        possible_lengths = (1, 4 * 1024, 32 * 1024, 64 * 1024)
+        retries_remaining = retry
+        length = len(data)
+        offset = 0
+        write_addr = addr + offset
+        with tqdm(desc="  Verifying", unit="B", unit_scale=True, total=length, disable=False) as pbar:
+            while length > 0 and retries_remaining > 0:
```
Comment (tinyfpga) [inline programmer/tinyprog/__init__.py]: Verify whether this entire while loop is necessary here.  It appears that this particular loop does nothing.

File: programmer/tinyprog/__init__.py [hunk 4]
```diff
@@ -161,7 +161,7 @@ def _read_metadata(self):
         import math
         meta_roots = (
             [self._parse_json(self.prog.read_security_register_page(p).replace(b"\x00", b"").replace(b"\xff", b"")) for p in [1, 2, 3]] +
-            [self._parse_json(self.prog.read(int(math.pow(2, p) - (4 * 1024)), (4 * 1024)).replace(b"\x00", b"").replace(b"\xff", b"")) for p in [17, 18, 19, 20, 21, 22, 23, 24]]
+            [self._parse_json(self.prog.read(int(math.pow(2, p) - (4 * 1024)), (4 * 1024) - 256).replace(b"\x00", b"").replace(b"\xff", b"")) for p in [17, 18, 19, 20, 21, 22, 23, 24]]
```
Comment (tinyfpga) [inline programmer/tinyprog/__init__.py]: Why did you change this to only read 256 bytes from the latest 4KB sector? The original code reads the full 4KB in order to give more space for the JSON metadata.

File: programmer/tinyprog/__init__.py [hunk 5]
```diff
@@ -376,6 +379,84 @@ def program(self, addr, data):
             self.progress("Failure!")
             return False
 
+    # for each sector: erase,program,verify
+    # this programming is much more reliable than previous program()
+    # because each written sector will be verified and in case of
+    # error, retried several times
+    def program_sectors(self, addr, data, retry=10):
```
Comment (tinyfpga) [inline programmer/tinyprog/__init__.py]: How often does the flash write fail and need to be retried?  I haven't experienced this before.

File: programmer/tinyprog/__init__.py [hunk 6]
```diff
@@ -376,6 +379,84 @@ def program(self, addr, data):
...
+        possible_lengths = (1, 4 * 1024, 32 * 1024, 64 * 1024)
+        retries_remaining = retry
+        length = len(data)
+        offset = 0
+        write_addr = addr + offset
+        with tqdm(desc="    Writing", unit="B", unit_scale=True, total=length, disable=False) as pbar:
+            while length > 0 and retries_remaining > 0:
```
Comment (tinyfpga) [inline programmer/tinyprog/__init__.py]: This entire while loop was copied and pasted from the `erase` function...instead of copy and pasting the entire body of the function, why not call the function from here instead?  I can't accept the same large block of code duplicated three times.

File: programmer/tinyprog/__main__.py [hunk 7]
```diff
@@ -331,7 +331,7 @@ def progress(info):
                 if not fpga.program_bitstream(addr, bitstream):
                     sys.exit(1)
 
-            fpga.boot()
+            # fpga.boot()
```
Comment (tinyfpga) [inline programmer/tinyprog/__main__.py]: Why is this commented out?  The programmer is supposed to reboot the target FPGA board into the user design after successfully programming the user bitstream.

File: programmer/initialize/boardmeta4MB.bin [hunk 8]
```diff
@@ -0,0 +1,119 @@
+{
```
Comment (tinyfpga) [inline programmer/initialize/boardmeta4MB.bin]: This file is specific to the ULX3S board, it doesn't belong in a generic path.  Instead it should go to the board-specific location of the repository, something like: `boards/ULX3S/initialize`

File: programmer/initialize/boardmeta8MB.bin [hunk 9]
```diff
@@ -0,0 +1,119 @@
+{
```
Comment (tinyfpga) [inline programmer/initialize/boardmeta8MB.bin]: This file is specific to the ULX3S board, it doesn't belong in a generic path.  Instead it should go to the board-specific location of the repository, something like: `boards/ULX3S/initialize`

File: programmer/initialize/initialize4MB.sh [hunk 10]
```diff
@@ -0,0 +1,28 @@
+#!/bin/sh -e
```
Comment (tinyfpga) [inline programmer/initialize/initialize4MB.sh]: This file is specific to the ULX3S board, it doesn't belong in a generic path.  Instead it should go to the board-specific location of the repository, something like: `boards/ULX3S/initialize`

File: programmer/initialize/initialize8MB.sh [hunk 11]
```diff
@@ -0,0 +1,29 @@
+#!/bin/sh -e
```
Comment (tinyfpga) [inline programmer/initialize/initialize8MB.sh]: This file is specific to the ULX3S board, it doesn't belong in a generic path.  Instead it should go to the board-specific location of the repository, something like: `boards/ULX3S/initialize`

File: programmer/tinyprog/__init__.py (hunk 3 above)
Comment (emard) [reply programmer/tinyprog/__init__.py]: I wanted to have double verify, first verify is sector-by-sector based but
final verify is against complete data. If e.g. sector address was wrong
or overlap for whatever reason, then each sector will verify OK but final
result won't be OK.

File: boards/ulx3s-v1.7-45f/top/bootloader_ulx3s.v [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+  ////////////////////////////////////////////////////////////////////////////////
+  ////////////////////////////////////////////////////////////////////////////////
+  wire reset;
+  wire usb_p_tx;
+  wire usb_n_tx;
+  wire usb_p_rx;
+  wire usb_n_rx;
+  wire usb_tx_en;
+  wire pin_led;
+  wire boot;
```
Comment (tinyfpga) [inline boards/ulx3s-v1.7-45f/top/bootloader_ulx3s.v]: Boot should be tied to a pin that is connected to PROGRAMN.  This is how I do it on the latest TinyFPGA EX prototypes.  Once boot is asserted it triggers the reconfigure process in the FPGA and it automatically boots the user design. 
image

File: programmer/tinyprog/__init__.py (hunk 4 above)
Comment (emard) [reply programmer/tinyprog/__init__.py]: When I initialize flash using scripts in "initialize" directory and if
I have jump command which must start at last 256 bytes, then it
makes no sense to concatenate this to JSON string.

additional 256 binary of jump command concatenated to JSON
makes my python exit with error :)

File: common/tinyfpga_bootloader.v [hunk 13]
```diff
@@ -74,7 +74,11 @@ module tinyfpga_bootloader (
       end
     end
   end
-  always @(posedge clk_48mhz) pwm_cnt <= pwm_cnt + 1'b1; 
+  always @(posedge clk_48mhz)
```
Comment (tinyfpga) [inline common/tinyfpga_bootloader.v]: It's unnecessary to reset the `pwm_cnt` register.  Since it is unnecessary I would rather keep it free-running without a reset.

File: programmer/tinyprog/__init__.py (hunk 5 above)
Comment (emard) [reply programmer/tinyprog/__init__.py]: I have 2 USB ports and 2 types of FLASH, spansion and issi (was micron).
spansion flash works good with both ports, retry 1x now and then.
issi on one port retries very much on particular places of long uploads.
5 retries is not enough 10 is makes it work. It's not problem with the flash
but with unreliable USB implementation that looses packets.

Same binary content makes same amount of retries on exactly the same
place in the data every time, no randomness therre.

File: programmer/tinyprog/__init__.py (hunk 3 above)
Comment (tinyfpga) [reply programmer/tinyprog/__init__.py]: Take a closer look at the while loop line by line...it's not actually doing anything.  It just updates local variables and updates the progress bar.
```python
            while length > 0 and retries_remaining > 0:
                erase_length = max(p for p in possible_lengths
                                   if p <= length and write_addr % p == 0)
                if erase_length == 1:
                    # there are no opcode to erase that much
                    # either we want to erase up to multiple of 0x1000
                    # or we want to erase up to length

                    # start_addr                            end_addr
                    # v                                     v
                    # +------------------+------------------+----------------+
                    # |       keep       |      erase       |      keep      |
                    # +------------------+------------------+----------------+
                    #  <- start_length -> <- erase_length -> <- end_length ->

                    start_addr = write_addr & 0xfff000
                    start_length = write_addr & 0xfff
                    erase_length = min(0x1000 - start_length, length)
                    end_addr = start_addr + start_length + erase_length
                    end_length = start_addr + 0x1000 - end_addr

                self.progress(erase_length)

                write_data = data[offset : offset + erase_length]
                # read_back = self.read(write_addr, erase_length)
                read_back = write_data

                # update
                if read_back == write_data:
                  length -= erase_length
                  write_addr += erase_length
                  offset += erase_length
                  retries_remaining = retry
                  pbar.update(erase_length)
                else:
                  retries_remaining -= 1
```

File: programmer/tinyprog/__init__.py (hunk 6 above)
Comment (emard) [reply programmer/tinyprog/__init__.py]: Yes actually the erase function was used as basis of sector program function.
I'm not a good python programmer, I needed quick fix just to make it work for me.
If someone can cleaner rewrite everything from scratch, its OK for me just
follow the logic: each sector should be erased, written and verified with multiple
retrieed (10x) for successful upload of 1-2MB or longer data 

I did similar bootloadings with sectors for AVR with unreliable
data link so I know what and how should be done.

Of course, if USB compatibility part is fixed to not loose packets,
then everyting will work even better. I suggest moving to HID packets
instead of trying to emulate usb-serial.

File: programmer/tinyprog/__init__.py (hunk 4 above)
Comment (tinyfpga) [reply programmer/tinyprog/__init__.py]: Ok, so it is keeping 4KB minus the 256 bytes required for the jump code.  Now that makes sense to me.

File: programmer/tinyprog/__init__.py (hunk 5 above)
Comment (tinyfpga) [reply programmer/tinyprog/__init__.py]: The behavior is different between the different flash chips? If its an issue with USB I would expect the behavior to be exactly the same between the chips.  If you tried flashing the same bitstream file to both chips and observe differences in retries?

File: programmer/tinyprog/__main__.py (hunk 7 above)
Comment (emard) [reply programmer/tinyprog/__main__.py]: It's just commented out because I don't want to leave bootloader after each
upload, e.g. for initialization I need to upload 3-4 times to different flash address
primariy image, golden image, jump commands, metadata....

If boot is executed after first upload, then this procedure can't be
conveniently automated in a script and user will get confused how to
initialize.

File: programmer/tinyprog/__init__.py (hunk 6 above)
Comment (tinyfpga) [reply programmer/tinyprog/__init__.py]: Whether it's HID or serial doesn't actually make a difference for the reliability.  In the end it's still two bulk endpoints transfering data.  There is likely a bug in the code, I have noticed some specific data patterns that will cause the issue, but I haven't been able to track it down.  I can see if I can't write up a smaller algorithm to do the retries.

How big of a sector are you using? 1MByte?

File: programmer/tinyprog/__main__.py (hunk 7 above)
Comment (tinyfpga) [reply programmer/tinyprog/__main__.py]: Ok, so the solution here is to add a command-line flag to disable automatic reboot after programming a user bitstream.  I would call it `--no-boot`.

File: programmer/initialize/boardmeta4MB.bin (hunk 8 above)
Comment (emard) [reply programmer/initialize/boardmeta4MB.bin]: it contains string ULX3S I'm sorry so It can go to my board area,
but I suggest that you place some generic initialization of your own,

I didn't find any so put something as better than nothing

File: programmer/initialize/boardmeta8MB.bin (hunk 9 above)
Comment (emard) [reply programmer/initialize/boardmeta8MB.bin]: ok

File: programmer/initialize/initialize4MB.sh (hunk 10 above)
Comment (emard) [reply programmer/initialize/initialize4MB.sh]: yes

File: programmer/initialize/jump.py [hunk 14]
```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/env python
```
[the remark this answers is not in the document]
Comment (emard) [reply programmer/initialize/jump.py]: It's excellent but it's not enough :(
This is "golden" jump command, but there exists
also a hidden form of "multi-boot" jump command which
diamond somehow patches to actual bitstream which I don't
know format of this.

[the remark this answers is not in the document]
Comment (emard) [reply programmer/initialize/jump.py]: jump.py is easy to use, see the initialize*.sh scripts

it takes address of golden image at commandline argument
and creates some binary which should be written to last 256 bytes

This binary will do golden jump, if primary image is damaged the
golden will be jumped to

File: common/tinyfpga_bootloader.v (hunk 13 above)
Comment (emard) [reply common/tinyfpga_bootloader.v]: Yes but for me it helps to visualise reset when it happens, the rhythm of the
fade-in-out LED will abruptly change so I know reset has happened

Comment (tinyfpga) [reply common/tinyfpga_bootloader.v]: I see.  That's a good reason.  I think it would be nice in the future to have some more states for the PWM LED to communicate some more.

File: programmer/tinyprog/__init__.py (hunk 3 above)
Comment (emard) [reply programmer/tinyprog/__init__.py]: Maybe you are right :) I had to uncomment
# read_back = self.read.bla
Let's say a verify-only commandline option would be good to have

File: programmer/tinyprog/__init__.py (hunk 4 above)
Comment (emard) [reply programmer/tinyprog/__init__.py]: Yes, yes, golden jump code must be at last 256 bytes, its location is hardcoded by lattice

File: programmer/tinyprog/__init__.py (hunk 5 above)
Comment (emard) [reply programmer/tinyprog/__init__.py]: The behaviour is differerent to different flash chips but my theory is that
this is only because different flash chips have different timings during waiting
for a sector to be erased or written and it causes USB code to be caught in
some unsupported state (bug exposed)

Same data content flashed to differrent chips behaves different in terms of retries.

But always in the same data on the same chip fail in predictable manner each time,
so my theory is that it's not e.g. worn-out flash memory or undervoltage supply which
causes data to be unpredictably written. Both chips are new, good, 100% functional and properly powered. If flashed with other tools each flash succeeds reliable and no verify
fails no retries needed.

File: programmer/tinyprog/__init__.py (hunk 6 above)
Comment (emard) [reply programmer/tinyprog/__init__.py]: There is (I guess) important but invisilble difference. Serial transfer
need to be bidirectional and each direction needs to be a part of continuous
data stream. My theory is that in verilog code of simplified USB protocol this
can't be properly handing without introducing much complexity with semaphores
buffering mutexing or similar. Simultaneously running bidirectional strems poses
additional challenge for potentioanly as-simple-as possible USB core.

HID doesn't treat data as streams on core level but as packets so it's more closer to UDP.
Assembly of the bidirectionaly stream will be left to python which is much more convenient
to be done than in verilog.

I'm using sector size same as erase size because the funciton is modified erase function.
so sector is most of the times 64K but at the end of data it usually becomes shorter.

File: programmer/tinyprog/__main__.py (hunk 7 above)
Comment (emard) [reply programmer/tinyprog/__main__.py]: yes, that's the good solution
